node_shell.py

```diff
@@ -10,74 +10,241 @@
 More details in the README.md file.
 
 Author: Giovanni Pizzi, EPFL
+Contributor: Bonan Zhu
 
 Version: 0.1
 """
 import cmd2
+from cmd2 import ansi
 import functools
 from datetime import datetime
 import pytz
 import sys
+import re
+from traceback import print_exc
+from pprint import pformat
+from collections import namedtuple
+import contextlib
 
 from ago import human
 import click
 
 from aiida.common.links import LinkType
+from aiida.cmdline.commands.cmd_verdi import verdi
 from aiida.orm.utils.repository import FileType
-from pprint import pformat
-
+from aiida.orm import GroupTypeString
 # Examples for autocompletion: https://github.com/python-cmd2/cmd2/blob/master/examples/tab_autocompletion.py
 
-LINK_TYPES_DICT = {link_type.value.upper(): link_type for link_type in LinkType}
+LINK_TYPES_DICT = {
+    link_type.value.upper(): link_type
+    for link_type in LinkType
+}
 
 LINK_TYPES = sorted(LINK_TYPES_DICT.keys())
 
+
 def with_default_argparse(f):
     """Decorator to use the default argparse parser (e.g. provide -h)."""
     return cmd2.with_argparser(cmd2.Cmd2ArgumentParser())(f)
 
+
 def now_aware():
     """Get the current datetime object, aware of the timezone (in UTC)."""
     return datetime.utcnow().replace(tzinfo=pytz.utc)
 
+
 def needs_node(f):
     """Decorator to add to a command if it needs to have a node loaded to work.
     
     It performs validation that the node is there, and prints a suitable error if this is not the case."""
     @functools.wraps(f)
-    def wrapper(self, *args, **kwds): # pylint: disable=inconsistent-return-statements
+    def wrapper(self, *args, **kwds):  # pylint: disable=inconsistent-return-statements
         """Print an error and return if there is no node loaded."""
         if not self._current_node:  # pylint: disable=protected-access
-            print("ERROR: No node loaded - load a node with `load NODE_IDENTIFIER` first")
+            self.poutput(
+                "ERROR: No node loaded - load a node with `load NODE_IDENTIFIER` first"
+            )
             return
         return f(self, *args, **kwds)
 
     return wrapper
 
 
+def needs_new_node(f):
```

Comment:
Maybe we want to put a `always=True` to always reload the node, for those properties that are mutable? Label, description, mtime, comments, extras.
And then put this extra, optional flag in the decorators for these properties?